### PR TITLE
[SPARK-25751][K8S][TEST] Unit Testing for Kerberos Support

### PR DESCRIPTION
**Pull request by ifilonenko**
## What changes were proposed in this pull request?

Unit tests for Kerberos support addition

## How was this patch tested?

Unit and Integration tests

### CODE REVIEW

**Comment by ifilonenko** [comment]
@liyinan926 @mccheah for review on unit testing

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesConf.scala [hunk 1]
```diff
@@ -78,6 +79,12 @@ private[spark] case class KubernetesConf[T <: KubernetesRoleSpecificConf](
 
   def krbConfigMapName: String = s"$appResourceNamePrefix-krb5-file"
 
+  // For unit-testing purposes
+  def hadoopBootstrapUtil: HadoopBootstrapUtil = new HadoopBootstrapUtil()
```
[the remark this answers is not in the document]
Comment (liyinan926) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/KubernetesConf.scala]: +1.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala [hunk 2]
```diff
@@ -63,4 +63,40 @@ object KubernetesFeaturesTestUtils {
...
   }
+
+  def podHasLabels(pod: Pod, labels: Map[String, String]): Boolean = {
+    labels.toSet.subsetOf(pod.getMetadata.getLabels.asScala.toSet)
+  }
+
+  // Kerberos Specific Test utils
+
+  // Upon use of bootstrapHadoopConfDir
+  def hConfBootPod(inputPod: SparkPod): SparkPod =
```
Comment (liyinan926) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala]: Can we just name it `hadoopConfBootPod`?

**Review by liyinan926 — CHANGES_REQUESTED** [review]

**Comment by ifilonenko** [comment]
retest this please

**Comment by ifilonenko** [comment]
retest this please

**Comment by ifilonenko** [comment]
retest this please

**Comment by mccheah** [comment]
I know that `HadoopBootstrapUtil` is being reworked soon - is it worth adding a unit test for that class here? @vanzin

**Comment by vanzin** [comment]
I'd say don't worry about any cleanup I'll be doing later - I'll take care of things at that point, updating tests as needed.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+      .withNewMetadata()
+      .withName(configMapName)
+      .endMetadata()
+      .addToData(Map(tmpFile.toPath.getFileName.toString -> "contents").asJava)
+      .build())
+  }
+
+
+  def createTempFile(dir: File, contents: String): File = {
+    val file = new File(dir, s"${UUID.randomUUID().toString}")
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala]: Use `File.createTempFile`.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        .endConfigMap()
+        .build(),
+      new VolumeBuilder()
+        .withName(SPARK_APP_HADOOP_SECRET_VOLUME_NAME)
+        .withNewSecret()
+          .withSecretName(dtSecretName)
+          .endSecret()
+        .build()
+    )
+    assert(podHasVolumes(resultingPod.pod, expectedVolumes))
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala]: If this assert (and others like it) fails, you'll get a pretty generic message ("false was not true" or some such).

Might be better to make `podHasVolumes` (and friends) assert with a more helpful message that shows at least the computed set.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    tmpDir = Utils.createTempDir()
+    tmpFile = createTempFile(tmpDir, "contents")
+  }
+
+  after {
+    tmpFile.delete()
+    tmpDir.delete()
+  }
+
+  test("Testing bootstrapKerberosPod with file location of krb5") {
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala]: Could you create better names for these tests? All of them start with "Testing", which is  redundant and also makes it harder to differentiate them.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    assert(resultingCMap === new ConfigMapBuilder()
+      .withNewMetadata()
+      .withName(configMapName)
+      .endMetadata()
+      .addToData(Map(tmpFile.toPath.getFileName.toString -> "contents").asJava)
+      .build())
+  }
+
+
+  def createTempFile(dir: File, contents: String): File = {
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala]: This method is called in a single place. Just inline it.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    val configMapName = "hconfMapName"
+    val resultingCMap = hadoopBootstrapUtil.buildkrb5ConfigMap(
+      configMapName,
+      tmpFile.getAbsolutePath
+    )
+    assert(resultingCMap === new ConfigMapBuilder()
+      .withNewMetadata()
+        .withName(configMapName)
+        .endMetadata()
+      .addToData(Map(tmpFile.toPath.getFileName.toString -> "contents").asJava)
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala]: all that long call is just `tmpFile.getName()`.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    val configMapName = "hconfMapName"
+    val resultingCMap = hadoopBootstrapUtil.buildHadoopConfigMap(
+      configMapName,
+      Seq(tmpFile)
+    )
+    assert(resultingCMap === new ConfigMapBuilder()
+      .withNewMetadata()
+      .withName(configMapName)
+      .endMetadata()
+      .addToData(Map(tmpFile.toPath.getFileName.toString -> "contents").asJava)
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala]: Same.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala [hunk 9]
```diff
@@ -63,4 +63,54 @@ object KubernetesFeaturesTestUtils {
   def containerHasEnvVar(container: Container, envVarName: String): Boolean = {
     container.getEnv.asScala.exists(envVar => envVar.getName == envVarName)
   }
+
+  def containerHasEnvVars(container: Container, envs: Map[String, String]): Boolean = {
+    envs.toSet.subsetOf(container.getEnv.asScala
+      .map { e => (e.getName, e.getValue)}.toSet)
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala]: space before `}`

Also in other places.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala [hunk 10]
```diff
@@ -63,4 +63,54 @@ object KubernetesFeaturesTestUtils {
...
+
+  def podHasLabels(pod: Pod, labels: Map[String, String]): Boolean = {
+    labels.toSet.subsetOf(pod.getMetadata.getLabels.asScala.toSet)
+  }
+
+  def podHasVolumes(pod: Pod, volumes: Seq[Volume]): Boolean = {
+    volumes.toSet.subsetOf(pod.getSpec.getVolumes.asScala.toSet)
+  }
+
+  // Kerberos Specific Test utils
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala]: Unnecessary comment.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala [hunk 11]
```diff
@@ -63,4 +63,54 @@ object KubernetesFeaturesTestUtils {
...
+
+  // Upon use of bootstrapHadoopConfDir
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala]: Not sure I understand what this comment means.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+      ENV_SPARK_USER -> userName)
+    )
+    containerHasVolumeMounts(resultingPod.container, Map(
+      SPARK_APP_HADOOP_SECRET_VOLUME_NAME -> SPARK_APP_HADOOP_CREDENTIALS_BASE_DIR,
+      KRB_FILE_VOLUME -> (KRB_FILE_DIR_PATH + "/krb5.conf"))
+    )
+  }
+
+  test("bootstrapKerberosPod with pre-existing configMap specified for krb5.conf file") {
+    val dtSecretName = "EXAMPLE_SECRET_NAME"
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala]: These can be constants at the top of the file or in a companion object.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    Files.write("contents".getBytes, tmpFile)
+  }
+
+  after {
+    tmpFile.delete()
+    tmpDir.delete()
+  }
+
+  test("bootstrapKerberosPod with file location specified for krb5.conf file") {
+    val dtSecretName = "EXAMPLE_SECRET_NAME"
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/hadooputils/HadoopBootstrapUtilSuite.scala]: These can be constants in a companion object or at the top of the class.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala [hunk 14]
```diff
@@ -63,4 +63,66 @@ object KubernetesFeaturesTestUtils {
...
+          .endMetadata()
+        .build(),
+      inputPod.container)
+
+  def subsetOfElem[T <: Set[B], B <: Any]: (T, T) => Boolean = (a, b) => a.subsetOf(b)
+  def subsetOfTup[T <: Set[(B, B)], B <: Any]: (T, T) => Boolean = (a, b) => a.subsetOf(b)
+
+  def assertHelper[T](con1: T, con2: T,
+      expr: (T, T) => Boolean = (a: T, b: T) => a == b, exprMsg: String = "equal to"): Unit = {
+    assert(expr(con1, con2), s"$con1 is not $exprMsg $con2 as expected")
```
Comment (mccheah) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala]: This interaction seems really awkward to me. Why can't we just embed the error message directly into the assertion on each assertion statement? I'm also not familiar enough with the ScalaTest framework but perhaps we can explore other options the library has to offer. (In Java I'm used to using [AssertJ](https://github.com/joel-costigliola/assertj-core) for fluent assertions but I'm not sure if that will be of any help for Scala objects, particularly Scala collections without converting everything.)

**Review by mccheah — CHANGES_REQUESTED** [review]

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala (hunk 14 above)
Comment (ifilonenko) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala]: I thought it would be better than doing a custom string at every assert statement. I don't find it to be too awkward, but *shrug* :) I actually kind of like it since we could do a check with any expression

Comment (mccheah) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/KubernetesFeaturesTestUtils.scala]: I think here the simpler approach is preferred - on each assertion line I want to see right then and there what error message I'll get back, not having to go through a helper method.

**Comment by ifilonenko** [comment]
gentle ping @mccheah :)

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    )).thenAnswer(new Answer[SparkPod] {
+      override def answer(invocation: InvocationOnMock) : SparkPod = {
+        KubernetesFeaturesTestUtils.hadoopConfBootPod(
+          invocation.getArgumentAt(3, classOf[SparkPod]))
+      }
+    })
+    val hConfStep = new HadoopConfExecutorFeatureStep(kubernetesConf, hadoopBootstrapUtil)
+    val pod = hConfStep.configurePod(sparkPod)
+    KubernetesFeaturesTestUtils.podHasLabels(pod.pod, Map("bootstrap-hconf" -> "true"))
+    assert(hConfStep.getAdditionalPodSystemProperties() == Map.empty,
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala]: use `assert(foo.isEmpty)`, no need for a message. Also in a bunch of places below.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+  before {
+    MockitoAnnotations.initMocks(this)
+    val sparkConf = new SparkConf(false)
+      .set(HADOOP_CONFIG_MAP_NAME, hadoopConfMapName)
+    when(kubernetesConf.sparkConf).thenReturn(sparkConf)
+  }
+
+  test("bootstrapHadoopConf being applied to a base spark pod") {
+    when(hadoopBootstrapUtil.bootstrapHadoopConfDir(
```
Comment (vanzin) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala]: So after I spent some time cleaning other code in the k8s backend, I started to really dislike this kind of test that over-mocks things. This test is actually just a result of a lot of other things that I think are wrong in the current code (not just kerberos-related, but in the k8s backend in general).

Basically you have a lot of code and mocks here to test a single thing: that the feature step is calling this one method with the parameters you expect.

If instead the feature step was written to be more easily testable, this wouldn't be needed. You wouldn't need to mock anything, and it would just run the normal code and you'd assert that the resulting pod has what you expect.

I filed SPARK-25874 to fix all this stuff (with a few sub-tasks), and I'm now thinking it's better to do that first. It would simplify the tests here a lot. It would avoid all this noisy and IMO unnecessary mocking. And we'd have better tests. You can look at my p.o.c. code linked from that bug to see what I mean.

As another data point, to address the comments on SPARK-25815 I'm having to rewrite a bunch of code in this area. So all these tests would have to be largely rewritten (on which PR, it depends on which one goes in first).

So, long story short, checking this in would mean I'd be rewriting all this code very, very soon. So I'm not seeing a lot of gains here...

**Review by vanzin — COMMENTED** [review]
See long comment.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala (hunk 16 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala]: Instead of mocks would we implement stub interfaces then? The purpose of the mocks is to make every unit test run only code that is in the class under test, plus utility methods. I think the only alternative to mocks is writing stub interfaces, and it's not clear how much more / less boilerplate that would be also. If the stub interfaces lead to less boilerplate than mocks then we should go with that.

Comment (vanzin) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala]: > The purpose of the mocks is to make every unit test run only code that is in the class under test

First, let me clarify that I'm against what I think is the excessive use of mocks in these tests; and by that I mean mostly mocking a whole bunch of things using mockito.

You can mock things without mockito. e.g. if what you want is to provide the test with a specific mock configuration, it's pretty trivial to do that with existing classes. Tests do this all over the code base without having to mock `SparkConf` or any other classes. 

(This is currently more painful than it should be in the k8s backend, which is one of the things I'm trying to fix with my cleanups.)

So let's use this particular test here as an example. What is it testing?

I think it should be testing that if a user has set the appropriate configuration in their `SparkConf`, that this step will mount a volume with the user's Hadoop configuration in the final pod. Is that what the code here is testing?

I don't think so. It's testing that a specific method call is being made with certain parameters. A much better test would be to actually let the code in the step run and do its thing, and test the final result for expected things (in this case, the hadoop conf being mounted).

Or thinking another way: how can you actually break this test, without removing that one method call from the step implementation? Since both the inputs and the output of that call are mocked here.

Comment (mccheah) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala]: There is a point of over-mocking, I agree. I think `KubernetesConf` shouldn't be mocked in most cases because it's just a structure. This is similar to not mocking something like a case class or a hash map. I also don't think we need to use Mockito Answers here - that could be done with stub implementations of the submodule. I didn't have a strong enough conviction of not using Mockito answers, but in general I think we should favor stub implementations over `Answer`; it's more readable.

I think it's fine though that we want to test "We're calling this submodule with these parameters", because the rest of the module's correctness is unit-test covered in the unit tests of that submodule. The general premise we'd like to follow is that a unit test should only execute code that is in the class under test. In other words, in this concrete case, since `HadoopBootstrapUtil` is a separate class, no code in `HadoopBootstrapUtil` should be run as part of the test. The class under test is responsible for calling the utility submodule with certain arguments but is not concerned about what that submodule actually does. Thus the test also doesn't have to be concerned here about what that submodule actually does, but should check that the submodule was actually called. If we want to check the submodule's correctness, we unit test the submodule.

Now - did we need a submodule to begin with? We could have, for example, kept HadoopBootstrapUtil not really be a submodule at all, but just a `static` call on a utility method. I think that's debatable - it makes each test need to cover a larger amount of code; that is, we no longer test the utility method in isolation, which is what we get here. Therefore it's unclear if HadoopBootstrapUtil being its own object that can be stubbed, is the right approach.

Comment (vanzin) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/features/HadoopConfExecutorFeatureStepSuite.scala]: > I think it's fine though that we want to test "We're calling this submodule with these parameters"

I think that's a very restrictive view of testing. I prefer my previous interpretation: the goal of this step is to mount a volume if a certain config is set. Whether it does that by calling into another module or not, it doesn't matter. That's what should be tested. You're taking "white box testing" to an extreme here - you're testing specific lines of code in the implementation instead of testing the contract ("for these inputs generate these outputs").

If it's done via another module, and you want to have unit tests for that module too, that's fine. But  then you get into your other question...

Looking from the other side: if you change the implementation in a way that breaks this test, but keeps the actual functionality, what good was this test doing?

> did we need a submodule to begin with?

My answer to that question is no - and as part of my updates to SPARK-25815 I'll be removing that class. But that's kinda orthogonal to my point.

**Comment by vanzin** [comment]
Do these tests still make sense after recent changes?

**Comment by vanzin** [comment]
These are now very out of date given changes in the feature code. If there's any missing test it would be a completely different PR.
